docs: copy best practices from Juju docs (#2203)
canonical/charmcraft: docs/reference/files/charmcraft-yaml-file.rst

        :start-at: actions:
        :end-before: analysis:

.. admonition:: Best practice

>     :class: hint

Reviewer: I don't think we need the admonitions. They feel like visual noise to my eyes.

Author: There are 3 reasons for these:
* We want to really call attention to these, because they are the markers of what we consider mature/quality charms.
* We want to have a common look for best practice across the charming ecosystem, so that even though the docs are decentralised they feel like parts of a common platform.
* We want to have the best practices in the places that they'll most likely be found by charmers, rather than in one big list like Juju used to have, but we also need to be able to automatically collect them all (for example, for checking in a quality review process). For this, it's not important what they look like, but collecting them across all the tools is much simpler if there's a standard (or rather two: one for Markdown and one for ReST) rather than having to implement collectors for each tool.

>     Prefer lowercase alphanumeric names, separated with dashes if required. For

Reviewer: ```suggestion
    Prefer lowercase alphanumeric names, and use hyphens (-) to separate words. For
```

        :start-at: config:
        :end-before: containers:

.. admonition:: Best Practice
    :class: hint

>     Charms should keep the user experience of the operator as simple and obvious

Reviewer 2: ```suggestion
    Charms should keep the user experience of the operator as smooth
```

    as possible. Aim to make the application deployable without providing any
    further configuration options: ``juju deploy foo`` is preferable to
    ``juju deploy foo --config required-option=value``.

> 

Reviewer 2: You might streamline this as: To make the user's experience as smooth as possible, aim to...

.. admonition:: Best Practice
    :class: hint

    Consider which of your application's configuration options you should
    initially expose. If your application has many config options, it may be
    prudent to provide access to a select few, and add support for more as the

>     need arises.

Reviewer 2: It feels like the intent here is to say: Be thrifty with how many configs = knobs you expose. Just like Juju, a charm is supposed to be an opinionated tool. I'd rephrase something like that. (Because "initially expose" makes me wonder: Is it okay to end up a gazillion configs later on?)

.. admonition:: Best Practice
    :class: hint

    Prefer lowercase alphanumeric names, separated with dashes if required. For
    charms that have already standardised on underscores, it is not necessary to
    change them, and it is better to be consistent within a charm then to have
    some config names be dashed and some be underscored.

.. admonition:: Best Practice
    :class: hint

>     The description a string type. YAML supports two types of formats for that:

Reviewer 2: ```suggestion
    The description is a string type. YAML supports two types of formats for that:
```

Reviewer 2: I understand that this in the original Best practice doc this was about the config key. However, I feel like this is a comment on the description keys in charmcraft.yaml more generally, so many it's really a best practice at the level of the entire doc. Speaking of which, one more candidate for that top-level note might be https://discourse.charmhub.io/t/yaml-anchors-and-aliases/5683 , which we used to have in the charm SDK docs and I never ported into Charmcraft docs because I wasn't sure about the relevance. @lengau Can you please advise?

    ``block scalar`` and ``flow scalar``. Prefer the block style (using ``|``)
    to keep newlines. Using ``>`` will replace newlines with spaces and make
    the result harder to read on Charmhub.

.. tip::

  For very complex applications, consider providing "configuration profiles",
  which can group values for large configs together. For example,
  "profile: large" that tweaks multiple options under the hood to optimise for
  larger deployments, or "profile: ci" for limited resource usage during
  testing.

> 

Reviewer 2: @lengau Do we really want this?

Reviewer 3: I think this is fine to keep, though it's probably a best practice as well.

        :start-at: config:
        :end-before: containers:

.. admonition:: Best practice
    :class: hint

    Just like Juju, a charm is an opinionated tool. Configure the application
    with the best defaults (ideally the application is deployable without
    providing any configuration at deploy time), and only expose application
    configuration options when necessary.

.. admonition:: Best practice
    :class: hint

    Prefer lowercase alphanumeric names, separated with dashes if required. For
    charms that have already standardised on underscores, it is not necessary to
    change them, and it is better to be consistent within a charm then to have
    some config names be dashed and some be underscored.

.. admonition:: Best practice
    :class: hint

>     

Reviewer 3: ```suggestion

```

        :start-at: config:
        :end-before: containers:

.. admonition:: Best practice
    :class: hint

    Make the user experience of the operator as smooth as possible. The
    application should be deployable without providing any further configuration
    options: ``juju deploy foo`` is preferable to
    ``juju deploy foo --config required-option=value``.

.. admonition:: Best practice
    :class: hint

    Just like Juju, a charm is an opinionated tool. Configure the application
    with the best defaults, and only expose application configuration options

>     when necessary.

Reviewer: Do you think we could combine these points? They feel almost the same to me. I prefer the copy in the second paragraph, at any rate.

.. admonition:: Best practice
    :class: hint

    Prefer lowercase alphanumeric names, separated with dashes if required. For
    charms that have already standardised on underscores, it is not necessary to
    change them, and it is better to be consistent within a charm then to have
    some config names be dashed and some be underscored.

.. admonition:: Best practice
    :class: hint

    The description is a string type. YAML supports two types of formats for
    that: ``block scalar`` and ``flow scalar``. Prefer the block style
    (using ``|``) to keep newlines. Using ``>`` will replace newlines with

>     spaces and make the result harder to read on Charmhub.

Reviewer: I would cut this. It's more like a nice-to-know for a how-to guide than essential text in a reference, in my opinion.

> .. tip::

Reviewer: Can we make this a regular paragraph? :eyes:

>   For very complex applications, consider providing "configuration profiles",
>   which can group values for large configs together. For example,
>   "profile: large" that tweaks multiple options under the hood to optimise for
>   larger deployments, or "profile: ci" for limited resource usage during
>   testing.

Reviewer: ```suggestion
  For very complex applications, consider providing configuration profiles,
  which can group values for large configs together. For example,
  a ``profile: large`` that tweaks multiple options under the hood to optimize for
  larger deployments, or a ``profile: ci`` for limited resource usage during
  testing.
```

It's been slow going, but we have been creeping toward US English.

        :start-at: links:
        :end-before: name:

.. admonition:: Best Practice
    :class: hint

    Documentation links should apply to the charm, and not to the application that
    is being charmed. Assume that the user already has basic competency in the use

>     of the application.

Reviewer 2: Note for the future: We'll want to add one more thing to this note, that is, links to examples of nicely documented charms.

        :start-at: name:
        :end-before: parts:

.. admonition:: Best Practice
    :class: hint

    The name should be slug-oriented (ASCII lowercase letters, numbers, and
    hyphens) and follow the pattern ``<workload name in full>[<function>][-k8s]``.
    For example, ``argo-server-k8s``.

    Don't include an organisation or publisher in the name.

    Don't add an ``operator`` or ``charm`` prefix or suffix. For naming a
    repository, see :ref:`initialise-a-charm`.

> 

Reviewer 2: Thank you for including the negatives (which I forgot to include when I added this text originally).

        :start-at: name:
        :end-before: parts:

.. admonition:: Best practice
    :class: hint

    The name should be slug-oriented (ASCII lowercase letters, numbers, and
    hyphens) and follow the pattern ``<workload name in full>[<function>][-k8s]``.
    For example, ``argo-server-k8s``.

>     Include the '-k8s' suffix on all charms that run on a Kubernetes cloud,

Reviewer: ```suggestion
    Include the ``-k8s`` suffix on all charms that run on a Kubernetes cloud,
```

    unless the charm has no workload or you know that there will never be
    a machine version of the charm.

>     Don't include an organisation or publisher in the name.

Reviewer: ```suggestion
    Don't include an organization or publisher in the name.
```
